src/Interpreters/Context.cpp

```diff
@@ -1109,43 +1109,43 @@ void Context::setCurrentDatabase(const String & name)
     calculateAccessRights();
 }
 
-void Context::setCurrentQueryId(const String & query_id)
+/// Generate random UUID, but using lower quality RNG,
+///  because Poco::UUIDGenerator::generateRandom method is using /dev/random, that is very expensive.
+/// NOTE: Actually we don't need to use UUIDs for query identifiers.
+/// We could use any suitable string instead.
+String Context::generateQueryId()
 {
-    if (!client_info.current_query_id.empty())
-        throw Exception("Logical error: attempt to set query_id twice", ErrorCodes::LOGICAL_ERROR);
+    union
+    {
+        char bytes[16];
+        struct
+        {
+            UInt64 a;
+            UInt64 b;
+        } words;
+    } random;
 
-    String query_id_to_set = query_id;
+    random.words.a = thread_local_rng(); //-V656
+    random.words.b = thread_local_rng(); //-V656
 
-    if (query_id_to_set.empty())    /// If the user did not submit his query_id, then we generate it ourselves.
+    /// Use protected constructor.
+    struct QueryUUID : Poco::UUID
     {
-        /// Generate random UUID, but using lower quality RNG,
-        ///  because Poco::UUIDGenerator::generateRandom method is using /dev/random, that is very expensive.
-        /// NOTE: Actually we don't need to use UUIDs for query identifiers.
-        /// We could use any suitable string instead.
-
-        union
-        {
-            char bytes[16];
-            struct
-            {
-                UInt64 a;
-                UInt64 b;
-            } words;
-        } random;
+        QueryUUID(const char * bytes, Poco::UUID::Version version) : Poco::UUID(bytes, version) {}
+    };
 
-        random.words.a = thread_local_rng(); //-V656
-        random.words.b = thread_local_rng(); //-V656
+    return QueryUUID(random.bytes, Poco::UUID::UUID_RANDOM).toString();
+}
 
-        /// Use protected constructor.
-        struct QueryUUID : Poco::UUID
-        {
-            QueryUUID(const char * bytes, Poco::UUID::Version version)
-                : Poco::UUID(bytes, version) {}
-        };
+void Context::setCurrentQueryId(const String & query_id)
```

Comment:
Client also has Context. We can simply reuse this method to set or generate query id (for both cases: where there is query_id specified explicitly in command line and for the case when we have to generate it). And then we can just obtain query_id from Context.

Reply:
However, it'll break the previous assumption 
 https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/Context.cpp#L1115

Reply:
Ok, got it.